rfcs/0029-smart-contracts.md

```diff
@@ -0,0 +1,549 @@
+## Summary
+[summary]: #summary
+
+This document outlines implementation proposals for several versions of smart
+contracts that could be implemented on top of the coda blockchain.
+
+## Motivation
+[motivation]: #motivation
+
+These proposals are intended to extend the permissions model of accounts, so
+that transactions (or other state-change requests) may be allowed to apply to
+an account that the sender would usually not have permission to modify. These
+extended permissions are implemented as 'predicates', which the transaction
+must satisfy in order to be accepted and applied to the blockchain.
+
+#### Simple motivating example: transaction itermediary
+
+Alice wants to send a transaction to Bob, but has asked Claire to be her
+intermediary. Alice wants to ensure that
+- Claire is the only account that may issue a transaction on her account
+- Bob is the only account that Claire may issue transactions to
+- Claire may only issue a transaction up to some limit, or in some fixed
+  amount
+
+In this case, Alice should be able to add a predicate that guarantees these
+conditions, so that only transactions matching these may be accepted into the
+blockchain.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+NB: For the purposes of this description, the implementation and economics of
+on-chain data storage is treated as abstract. The assumptions are only that:
+* the data is associated with an account, which may be addressed through the
+  accounts Merkle tree
+* the data is stored in some cryptographically verifiable, addressible form.
+
+These proposals are approximately incremental: each proposal may be built out
+upon the foundation of the previous proposal.
+
+### Witness generation + proving off chain: store verification keys in accounts
```

Comment:
there are proof/verification keys generated for each account that wishes to run smart contracts?